File: easy/121. Best Time to Buy and Sell Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // 7 1 5 3 6 4
        //Time Complexity = O(n)
        //Space Complexity = O(1)
  
        int maxCur = 0;
        int result = 0;
        for(int i=1; i<prices.length; i++){
            maxCur = Math.max(0, maxCur + prices[i] - prices[i-1]);
            result = Math.max(result, maxCur);
        }
        return result;
    }
}
    

File: easy/121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        '''
        prices = [7,1,5,3,6,4] => 5
        prices = [7] => 0
        prices = [4,3,2,1] => 0
        
        TC:O(n)
        SC:O(1)
        '''
        
        min_price = prices[0]
        res = 0
        
        for price in prices :
            res = max(price - min_price, res)
            min_price = min(price, min_price)
            
        return res
